File: src-javascript/engine/platform/expansions.js
"use strict";

const EXPANSIONS_PATH = "/expansions";
const EXPANSIONS_SYMBOLIC_PATH = "/~expansions/";
const EXPANSIONS_FUNKIN_EXPANSION_NAME = "funkin";
const EXPANSIONS_FUNKIN_EXPANSION_PATH = EXPANSIONS_SYMBOLIC_PATH + EXPANSIONS_FUNKIN_EXPANSION_NAME;
const EXPANSIONS_CHAIN_FILENAME = "chain.ini";
const EXPANSIONS_ABOUT_FILENAME = "expansion.json";
const EXPANSIONS_COMMENT_CHAR = ';';
const EXPANSIONS_SELF_NAME = "/self/";

var expansions_chain_array = null;
var expansions_chain_array_size = 0;
var expansions_overrided_weeks_folder = null;


async function expansions_load(expansion_name) {
    if (expansion_name && fs_is_invalid_filename(expansion_name)) {
        console.error(`expansions_load() invalid expansion name, ilegal filename: ${expansion_name}`);
        return;
    }

    //for (let i = 0 ; i < expansions_chain_array_size ; i++) {
    //    if (expansions_chain_array[i] != Symbol(EXPANSIONS_FUNKIN_EXPANSION_PATH)) {
    //        expansions_chain_array[i] = undefined;
    //    }
    //}
    expansions_chain_array = undefined;
    expansions_chain_array_size = 0;

    // PreloadCache is not supported
    // force clear preload cache
    //PreloadCache.ClearCache();

    let chain = arraylist_init();

    if (await io_resource_exists(EXPANSIONS_FUNKIN_EXPANSION_PATH, false, true)) {
        await expansions_internal_load_dependency(chain, EXPANSIONS_FUNKIN_EXPANSION_NAME);
    }

    await expansions_internal_load_dependency(chain, expansion_name);

    if(expansions_overrided_weeks_folder) expansions_overrided_weeks_folder = undefined;
    expansions_overrided_weeks_folder = null;

    let expansion_base_path = string_concat(3, EXPANSIONS_SYMBOLIC_PATH, expansion_name, FS_CHAR_SEPARATOR);
    let about_path = string_concat(2, expansion_base_path, EXPANSIONS_ABOUT_FILENAME);
    let has_about = await io_resource_exists(about_path, true, false);

    if (has_about) {
        let about_json_path = string_concat(2, expansion_base_path, EXPANSIONS_ABOUT_FILENAME);

        about_path = undefined;
        about_path = await io_native_get_path(about_json_path, true, false, false);

        let json = await json_load_direct(about_path);
        let override_weeks_folder = json_read_boolean(json, "overrideWeeksFolder", false);

        // (JS & C#) only
        let window_icon = json_read_string(json, "windowIcon", null);
        let window_title = json_read_string(json, "windowTitle", null);
        await expansions_internal_update_window(expansion_base_path, window_title, window_icon);

        json_destroy(json);
        about_json_path = undefined;

        if (override_weeks_folder) {
            expansions_overrided_weeks_folder = string_concat(2, expansion_base_path, "weeks");
        }
    }

    expansion_base_path = undefined;
    about_path = undefined;

    expansions_chain_array_size = arraylist_trim(chain);
    expansions_chain_array = arraylist_peek_array(chain);
    arraylist_destroy(chain, true);
}

async function expansions_resolve_path(path, is_file, is_folder) {
    if (path == null) return null;
    if (!path.startsWith(FS_ASSETS_FOLDER, 0)) return path;

    let path_length = path.length;
    let index = FS_ASSETS_FOLDER.length;
    if (index < path_length && path[index] == FS_CHAR_SEPARATOR) index++;

    let relative_path = path.substring(index, path_length);
    let last_overrided_path = strdup(path);

    for (let i = 0; i < expansions_chain_array_size; i++) {
        let overrided_path = string_concat(
            3, expansions_chain_array[i], FS_CHAR_SEPARATOR, relative_path
        );

        if (await io_resource_exists(overrided_path, is_file, is_folder)) {
            if (last_overrided_path) last_overrided_path = undefined;
            last_overrided_path = overrided_path;
            continue;
        }

        overrided_path = undefined;
    }

    relative_path = undefined;

    return last_overrided_path;
}

async function expansions_get_path_from_expansion(path, expansion_index) {
    if (!path) return null;
    if (!path.startsWith(FS_ASSETS_FOLDER, 0)) return null;
    if (expansion_index < 0 || expansion_index >= expansions_chain_array_size) return null;

    let path_length = path.length;
    let index = FS_ASSETS_FOLDER.length;
    if (index < path_length && path[index] == FS_CHAR_SEPARATOR) index++;

    let relative_path = path.substring(index, path_length);

    let new_path = string_concat(
        3, expansions_chain_array[expansion_index], FS_CHAR_SEPARATOR, relative_path
    );

    relative_path = undefined;

    return new_path;

    /*if (await io_resource_exists(new_path, true, true)) {
        return new_path;
    }

    new_path = undefined;
    return null;*/
}


async function expansions_internal_load_dependency(chain, expansion_name) {
    if (!expansion_name) return;

    let expansion_path = string_concat(2, EXPANSIONS_SYMBOLIC_PATH, expansion_name);
    if (!await io_resource_exists(expansion_path, false, true)) {
        console.error(`expansions_load() '${expansion_name}' not found in ${expansion_path}`);
        expansion_path = undefined;
        return;
    }

    // PreloadCache is not supported
    //let expansion_preload_ini_path = string_concat(3, expansion_path, FS_CHAR_SEPARATOR, PRELOADCACHE_PRELOAD_FILENAME);
    //if (await io_resource_exists(expansion_preload_ini_path, true, false)) {
    //    PreloadCache.AddFileList(expansion_preload_ini_path);
    //}
    //expansion_preload_ini_path = undefined;

    let chain_ini_path = string_concat(3, expansion_path, FS_CHAR_SEPARATOR, EXPANSIONS_CHAIN_FILENAME);
    let has_chain = await io_resource_exists(chain_ini_path, true, false);

    if (!has_chain) {
        expansions_internal_add_to_chain(chain, expansion_path);
        expansion_path = undefined;
        chain_ini_path = undefined;
        return;
    }

    let unparsed_chain = await io_native_foreground_request(chain_ini_path, IO_REQUEST_TEXT);
    chain_ini_path = undefined;

    if (!unparsed_chain) return;

    let tokenizer = tokenizer_init("\r\n", true, false, unparsed_chain);
    if (!tokenizer) {
        expansions_internal_add_to_chain(chain, expansion_path);
        unparsed_chain = undefined;
        return;
    }

    let name;
    while ((name = tokenizer_read_next(tokenizer)) != null) {
        if (!name || name[0] == EXPANSIONS_COMMENT_CHAR) {
            name = undefined;
            continue;
        }

        if (name == EXPANSIONS_SELF_NAME) {
            expansions_internal_add_to_chain(chain, expansion_path);
        } else if (fs_is_invalid_filename(name)) {
            console.error(`expansions_internal_load_dependency() '${name}' is not a valid folder name`);
        } else {
            // warning: recursive call
            await expansions_internal_load_dependency(chain, name);
        }

        name = undefined;
    }

    expansions_internal_add_to_chain(chain, expansion_path);

    tokenizer_destroy(tokenizer);
    unparsed_chain = undefined;
    tokenizer = undefined;
}

function expansions_internal_add_to_chain(chain, expansion_path) {
    let lowercase_expansion_name = string_to_lowercase(expansion_path);

    for (let expansion of arraylist_iterate4(chain)) {
        if (expansion == expansion_path) return;

        let lowercase_expansion = string_to_lowercase(expansion);
        let equals = lowercase_expansion == lowercase_expansion_name;

        lowercase_expansion = undefined;

        if (equals) {
            lowercase_expansion_name = undefined;
            return;
        }
    }

    lowercase_expansion_name = undefined;
    arraylist_add(chain, expansion_path);
}

async function expansions_internal_update_window(expansion_path, window_title, window_icon) {
    document.title = window_title ?? "";
    if ("main_layout_visor" in window) document.title = "[Layout visor] " + document.title;

    var /**@type {HTMLLinkElement} */ link = document.querySelector("link[rel~='icon']");

    if (!link) {
        if (!window_icon) return;
        link = document.createElement('link');
        link.rel = 'icon';
        document.head.appendChild(link);
    }

    if (!window_icon) {
        link.remove();
        return;
    }

    window_icon = fs_resolve_path(expansion_path + FS_CHAR_SEPARATOR + window_icon);
    link.href = await io_native_get_path(window_icon, true, false, false);
}

File: src-dreamcast/src/platform/expansions.c
#include "expansions.h"
#include "arraylist.h"
#include "fs.h"
#include "io.h"
#include "jsonparser.h"
#include "malloc_utils.h"
#include "preloadcache.h"
#include "stringutils.h"
#include "tokenizer.h"

const char* EXPANSIONS_PATH = "/expansions";
const char* EXPANSIONS_SYMBOLIC_PATH = "/~expansions/";
const char* EXPANSIONS_FUNKIN_EXPANSION_NAME = "funkin";
const char* EXPANSIONS_FUNKIN_EXPANSION_PATH = "/~expansions/funkin";
const char* EXPANSIONS_CHAIN_FILENAME = "chain.ini";
const char* EXPANSIONS_ABOUT_FILENAME = "expansion.json";
const char EXPANSIONS_COMMENT_CHAR = ';';
const char* EXPANSIONS_SELF_NAME = "/self/";

char** expansions_chain_array = NULL;
int32_t expansions_chain_array_size = 0;
char* expansions_overrided_weeks_folder = NULL;


void expansions_internal_load_dependency(ArrayList chain, const char* expansion_name);
void expansions_internal_add_to_chain(ArrayList chain, char* expansion_path);


void expansions_load(const char* expansion_name) {
    if (expansion_name && fs_is_invalid_filename(expansion_name)) {
        logger_error("expansions_load() invalid expansion name, ilegal filename: %s", expansion_name);
        return;
    }

    for (int32_t i = 0; i < expansions_chain_array_size; i++) {
        if (!string_equals(expansions_chain_array[i], EXPANSIONS_FUNKIN_EXPANSION_PATH)) {
            free_chk(expansions_chain_array[i]);
        }
    }
    free_chk(expansions_chain_array);
    expansions_chain_array_size = 0;

    // clear preload cache
    preloadcache_clear(false);

    ArrayList chain = arraylist_init(sizeof(char*));

    if (io_resource_exists(EXPANSIONS_FUNKIN_EXPANSION_PATH, false, true)) {
        expansions_internal_load_dependency(chain, EXPANSIONS_FUNKIN_EXPANSION_NAME);
    }

    expansions_internal_load_dependency(chain, expansion_name);

    if (expansions_overrided_weeks_folder) free_chk(expansions_overrided_weeks_folder);
    expansions_overrided_weeks_folder = NULL;

    char* expansion_base_path = string_concat(3, EXPANSIONS_SYMBOLIC_PATH, expansion_name, FS_STRING_SEPARATOR);
    char* about_path = string_concat(2, expansion_base_path, EXPANSIONS_ABOUT_FILENAME);
    bool has_about = io_resource_exists(about_path, true, false);

    if (has_about) {
        char* about_json_path = string_concat(2, expansion_base_path, EXPANSIONS_ABOUT_FILENAME);

        free_chk(about_path);
        about_path = io_get_native_path(about_json_path, true, false, false);

        JSONToken json = json_load_direct(about_path);
        bool override_weeks_folder = json_read_boolean(json, "overrideWeeksFolder", false);

        json_destroy(&json);
        free_chk(about_json_path);

        if (override_weeks_folder) {
            expansions_overrided_weeks_folder = string_concat(2, expansion_base_path, "weeks");
        }
    }

    free_chk(expansion_base_path);
    free_chk(about_path);

    arraylist_destroy2(&chain, &expansions_chain_array_size, (void**)&expansions_chain_array);
}

char* expansions_resolve_path(const char* path, bool is_file, bool is_folder) {
    if (path == NULL) return NULL;
    if (!string_starts_with(path, FS_ASSETS_FOLDER)) return string_duplicate(path);

    size_t path_length = strlen(path);
    int32_t index = string_get_length(FS_ASSETS_FOLDER);
    if (index < path_length && path[index] == FS_CHAR_SEPARATOR) index++;

    char* relative_path = string_substring(path, index, (int32_t)path_length);
    char* last_overrided_path = string_duplicate(path);

    for (int32_t i = 0; i < expansions_chain_array_size; i++) {
        char* overrided_path = string_concat(
            3, expansions_chain_array[i], FS_STRING_SEPARATOR, relative_path
        );

        if (io_resource_exists(overrided_path, is_file, is_folder)) {
            if (last_overrided_path) free_chk(last_overrided_path);
            last_overrided_path = overrided_path;
            continue;
        }

        free_chk(overrided_path);
    }

    free_chk(relative_path);

    return last_overrided_path;
}

char* expansions_get_path_from_expansion(const char* path, int32_t expansion_index) {
    if (!path) return NULL;
    if (!string_starts_with(path, FS_ASSETS_FOLDER)) return NULL;
    if (expansion_index < 0 || expansion_index >= expansions_chain_array_size) return NULL;

    size_t path_length = strlen(path);
    size_t index = strlen(FS_ASSETS_FOLDER);
    if (index < path_length && path[index] == FS_CHAR_SEPARATOR) index++;

    char* new_path = string_concat_with_substring(
        path, (int32_t)index, (int32_t)path_length, 2,
        2, expansions_chain_array[expansion_index], FS_STRING_SEPARATOR
    );

    return new_path;

    /*if (io_resource_exists(new_path, true, true)) {
        return new_path;
    }

    free_chk(new_path);
    return NULL;*/
}


void expansions_internal_load_dependency(ArrayList chain, const char* expansion_name) {
    if (!expansion_name) return;

    char* expansion_path = string_concat(2, EXPANSIONS_SYMBOLIC_PATH, expansion_name);
    if (!io_resource_exists(expansion_path, false, true)) {
        logger_error("expansions_load() '%s' not found in %s", expansion_name, expansion_path);
        free_chk(expansion_path);
        return;
    }

// not implemented
#if 0
    char* expansion_preload_ini_path = string_concat(3, expansion_path, FS_STRING_SEPARATOR, PRELOADCACHE_PRELOAD_FILENAME);
    if (io_resource_exists(expansion_preload_ini_path, true, false)) {
        preloadcache_add_filelist(expansion_preload_ini_path);
    }
    free_chk(expansion_preload_ini_path);
#endif

    char* chain_ini_path = string_concat(3, expansion_path, FS_STRING_SEPARATOR, EXPANSIONS_CHAIN_FILENAME);
    bool has_chain = io_resource_exists(chain_ini_path, true, false);

    if (!has_chain) {
        expansions_internal_add_to_chain(chain, expansion_path);
        free_chk(expansion_path);
        free_chk(chain_ini_path);
        return;
    }

    ArrayBuffer unparsed_chain = io_read_arraybuffer(chain_ini_path);
    free_chk(chain_ini_path);

    if (!unparsed_chain) return;

    Tokenizer tokenizer = tokenizer_init2("\r\n", true, (const char*)unparsed_chain->data, unparsed_chain->length);
    if (!tokenizer) {
        expansions_internal_add_to_chain(chain, expansion_path);
        free_chk(unparsed_chain);
        return;
    }

    char* name;
    while ((name = tokenizer_read_next(tokenizer)) != NULL) {
        if (!name || name[0] == EXPANSIONS_COMMENT_CHAR) {
            free_chk(name);
            continue;
        }

        if (string_equals(name, EXPANSIONS_SELF_NAME)) {
            expansions_internal_add_to_chain(chain, expansion_path);
        } else if (fs_is_invalid_filename(name)) {
            logger_error("expansions_internal_load_dependency() '%s' is not a valid folder name", name);
        } else {
            // warning: recursive call
            expansions_internal_load_dependency(chain, name);
        }

        free_chk(name);
    }

    expansions_internal_add_to_chain(chain, expansion_path);

    tokenizer_destroy(&tokenizer);
    free_chk(unparsed_chain);
    free_chk(tokenizer);
}

void expansions_internal_add_to_chain(ArrayList chain, char* expansion_path) {
    char* lowercase_expansion_name = string_to_lowercase(expansion_path);

    foreach (char*, expansion, ARRAYLIST_ITERATOR, chain) {
        if (string_equals(expansion, expansion_path)) return;

        char* lowercase_expansion = string_to_lowercase(expansion);
        bool equals = string_equals(lowercase_expansion, lowercase_expansion_name);

        free_chk(lowercase_expansion);

        if (equals) goto L_return;
    }
    arraylist_add(chain, expansion_path);

L_return:
    free_chk(lowercase_expansion_name);
}
